### ISSUE TITLE
Implement `migrate export` command

### ISSUE THREAD

**Issue by PastelMobileSuit** [git-lfs/git-lfs#3057]
We should add a new `export` subcommand to the `migrate` command. `migrate export` will do the opposite of `migrate import`. When given a range of references and a set of patterns, `migrate export` will rewrite git history within the given range to replace all LFS pointers matching the given patterns with the contents of the large files being pointed to.

By default, `migrate export` should require a reference range to work, but should also provide a `--everything` flag to allow rewriting of the entire history. We may also want to consider a `--no-rewrite` flag similar to that for the `migrate import` command, which would export a given set of files in a new commit without rewriting git history.

**Comment by ttaylorr** [comment]
Hi @PastelMobileSuit, this seems like a great idea to me, and I definitely am excited about seeing it land in Git LFS. Before diving into the details, I think that we should establish a good timeline for releasing this. Between the rest of the planning for v2.5.0 and the `--no-rewrite` changes you landed in #3029, I think that this is a good candidate for v2.6.0. I definitely don't want to rush you or shut you out of a release, so if you feel like v2.6.0 is a bad estimate, let me know. We can obviously re-evaluate as you work through this.

> We should add a new `export` subcommand to the `migrate` command. `migrate export` will do the opposite of `migrate import`.

OK, that sounds good. I would expect that for any patterns X, Y, and references Z, that the following would be an effective no-op:

```ShellSession
$ git lfs migrate import --include-ref=Z --include=X --exclude=Y
$ git lfs migrate export --include-ref=Z --include=X --exclude=Y
```

> When given a range of references and a set of patterns, `migrate export` will rewrite git history within the given range to replace all LFS pointers matching the given patterns with the contents of the large files being pointed to.

Sounds good. I think that there are two main conceptual challenges that I think of:

1. How do we download all of the objects that we need?
2. What is the reverse migration in `.gitattributes`?

For (1) I think that doing a scan of all history in the range of the migration (i.e., everything between `--include-ref` and `--exclude-ref`) fo all LFS pointers, and then sending those to the transfer queue (in package `tq`) for download. This has the drawback of potentially downloading a lot of data, but in practice I'm not so concerned as to pursue other approaches immediately. I think as long as we are good stewards of the cache (i.e., that we prune anything in `.git/lfs/objects` that isn't reachable any longer) that we should be OK. Definitely let me know if you feel differently.

For (2) I think we have a couple of approaches. The most obvious is to take a `.gitattributes` prior to migration, and change it as follows:

```ShellSession
$ cat .gitattributes
*.psd filter=lfs diff=lfs merge=lfs -text

$ git lfs migrate export --include="*.psd"
# ...

$ cat .gitattributes
*.psd filter=lfs diff=lfs merge=lfs -text
*.psd !filter !diff !merge +text
```

That is, for any new pattern that we get, immediately append it with `!filter` (read as, "unset the `filter` attribute for the given pattern") instead of `filter=lfs`. That works, but generates a redundant `.gitattributes`, when it might be cleaner to remove the `*.psd` line altogether.

One approach that we could take is to detect cases like this and simply remove the line. We'd have to make sure that that is a correct change first before doing so (i.e., that we haven't tracked or untracked any files that we didn't mean to) but I think that could be interesting.

Alternatively, I think that it would also be interesting to determine the set theoretic difference of two patterns (i.e., what's in `*.psd` and not in `**/foo/*.psd`?) This might be too much for v2.6.0, but these later approaches have the benefit of being able to be shipped later on, rather than right away.

**Comment by ttaylorr** [comment]
Great work on implementing this, @PastelMobileSuit! I think that we can close this because #3084 has merged. Let's keep this around for looking at some of the other issues in the future, when you have time. For now, I'll close this to mark it as 'done'.
